File: app/src/main/java/za/co/dvt/android/showcase/presentation/listapps/AppAdapter.java
package za.co.dvt.android.showcase.presentation.listapps;

import android.content.Context;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import java.util.List;

import za.co.dvt.android.showcase.R;
import za.co.dvt.android.showcase.model.AppModel;

public class AppAdapter extends RecyclerView.Adapter<AppViewHolder> {
    private final Context context;
    private List<AppModel> items;

    public AppAdapter(List<AppModel> items, Context context) {
        this.items = items;
        this.context = context;
    }

    @Override
    public AppViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        View v = LayoutInflater.from(parent.getContext()).inflate(R.layout.list_item_app, parent, false);
        return new AppViewHolder(v);
    }

    @Override
    public void onBindViewHolder(AppViewHolder holder, int position) {
        AppModel item = items.get(position);
        holder.textViewAppName.setText(item.name());

    }

    @Override
    public int getItemCount() {
        if (items == null) {
            return 0;
        }
        return items.size();
    }

    public void setItems(final List<AppModel> appModelList) {
        this.items = appModelList;
        notifyDataSetChanged();
    }
}

File: app/src/main/java/za/co/dvt/android/showcase/ui/listapps/AppAdapter.kt
package za.co.dvt.android.showcase.ui.listapps

import android.content.Context
import android.support.v7.widget.RecyclerView
import android.text.TextUtils
import android.view.LayoutInflater
import android.view.ViewGroup
import com.bumptech.glide.Glide
import com.google.firebase.storage.FirebaseStorage
import com.google.firebase.storage.StorageReference
import za.co.dvt.android.showcase.R
import za.co.dvt.android.showcase.model.AppModel
import za.co.dvt.android.showcase.utils.FirebaseImageLoader

class AppAdapter(private var items: List<AppModel>, private val context: Context, private val taskNavigator: ListAppsNavigator) : RecyclerView.Adapter<AppViewHolder>() {

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): AppViewHolder {
        val v = LayoutInflater.from(parent.context).inflate(R.layout.list_item_app, parent, false)
        return AppViewHolder(v)
    }

    override fun onBindViewHolder(holder: AppViewHolder, position: Int) {
        val appItem = items[position]
        holder.textViewAppName.text = appItem.name
        holder.textViewDescription.text = appItem.shortDescription
        holder.itemView.tag = items[position]
        holder.itemView.setOnClickListener { view ->
            val app = view.tag as AppModel
            taskNavigator.onAppClick(app)
        }
        val storageRef = appItem.getFullIconUrl()
        storageRef?.let{
            Glide.with(context).using(FirebaseImageLoader()).load(storageRef)
                    .into(holder.imageViewAppIcon)
        }
    }



    override fun getItemCount(): Int {
        return items.size
    }

    fun setItems(appModelList: List<AppModel>) {
        this.items = appModelList
        notifyDataSetChanged()
    }
}